### PR TITLE
feat: restore frame tree when reconnecting to browser

### PR DESCRIPTION
**Pull request by sadym-chromium**
Addressing https://github.com/GoogleChromeLabs/chromium-bidi/issues/2282

### CODE REVIEW

File: src/bidiMapper/modules/cdp/CdpTarget.ts [hunk 1]
```diff
@@ -133,9 +143,16 @@ export class CdpTarget {
    */
   async #unblock() {
     try {
+      await this.#cdpClient.sendCommand('Page.enable');
```
Comment (OrKoN) [inline src/bidiMapper/modules/cdp/CdpTarget.ts:146]: this is very likely not correct, you need to schedule the commands at the same time, as they might not reach the renderer while they are blocked, resulting in a deadlock

Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts]: right, fixed

File: src/bidiMapper/modules/cdp/CdpTarget.ts [hunk 2]
```diff
@@ -134,8 +145,17 @@ export class CdpTarget {
   async #unblock() {
     try {
       await Promise.all([
+        (async () => {
+          await this.#cdpClient.sendCommand('Page.enable');
+          // There can be some existing frames in the target, if reconnecting to an
+          // existing browser instance, e.g. via Puppeteer. Need to restore the browsing
+          // contexts for the frames to correctly handle further events, like
+          // `Runtime.executionContextCreated`.
+          const frameTree =
```
Comment (OrKoN) [inline src/bidiMapper/modules/cdp/CdpTarget.ts:154]: this is also incorrect: getFrameTree will only be sent once Page.enable is resolved, which is too late

**Comment by OrKoN** [comment]
@sadym-chromium this should probably work https://github.com/GoogleChromeLabs/chromium-bidi/pull/2290

File: src/bidiMapper/modules/cdp/CdpTarget.ts (hunk 2 above)
Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts]: ack

File: src/bidiMapper/modules/cdp/CdpTarget.ts [hunk 3]
```diff
@@ -134,8 +144,20 @@ export class CdpTarget {
...
         this.#cdpClient.sendCommand('Page.enable'),
+        // There can be some existing frames in the target, if reconnecting to an
+        // existing browser instance, e.g. via Puppeteer. Need to restore the browsing
+        // contexts for the frames to correctly handle further events, like
+        // `Runtime.executionContextCreated`.
+        // It's important to schedule this task together with enabling domains commands to
+        // prepare the tree before the events (e.g. Runtime.executionContextCreated) start
+        // coming.
+        this.#cdpClient
+          .sendCommand('Page.getFrameTree')
```
Comment (OrKoN) [inline src/bidiMapper/modules/cdp/CdpTarget.ts:156]: perhaps we could add a test to make sure that evaluate works right after a context creation (the puppeteer job does not report that scenario as fixed yet)

**Comment by OrKoN** [comment]
Let's link issues that this contributes to/closes

File: src/bidiMapper/modules/cdp/CdpTarget.ts (hunk 3 above)
Comment (sadym-chromium) [reply src/bidiMapper/modules/cdp/CdpTarget.ts:156]: The puppeteer tests started passing:
> **Remove** the following from the TestExpectations.json to ignore the error:
> * [launcher.spec] Launcher specs Puppeteer Puppeteer.connect should be able to reconnect to a disconnected browser"
> * [launcher.spec] Launcher specs Puppeteer Puppeteer.connect should be able to reconnect"

**Review by OrKoN — APPROVED** [review]

**Pull request merged**